extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/mail/GoogleMailExporter.java

```diff
@@ -0,0 +1,132 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dataportabilityproject.datatransfer.google.mail;
+
+import com.google.api.client.auth.oauth2.BearerToken;
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.services.gmail.Gmail;
+import com.google.api.services.gmail.Gmail.Users.Messages;
+import com.google.api.services.gmail.model.ListMessagesResponse;
+import com.google.api.services.gmail.model.Message;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.dataportabilityproject.datatransfer.google.common.GoogleStaticObjects;
+import org.dataportabilityproject.spi.transfer.provider.ExportResult;
+import org.dataportabilityproject.spi.transfer.provider.ExportResult.ResultType;
+import org.dataportabilityproject.spi.transfer.provider.Exporter;
+import org.dataportabilityproject.spi.transfer.types.ContinuationData;
+import org.dataportabilityproject.spi.transfer.types.ExportInformation;
+import org.dataportabilityproject.spi.transfer.types.IdOnlyContainerResource;
+import org.dataportabilityproject.spi.transfer.types.PaginationData;
+import org.dataportabilityproject.spi.transfer.types.StringPaginationToken;
+import org.dataportabilityproject.types.transfer.auth.TokensAndUrlAuthData;
+import org.dataportabilityproject.types.transfer.models.mail.MailContainerResource;
+import org.dataportabilityproject.types.transfer.models.mail.MailMessageModel;
+
+public class GoogleMailExporter implements Exporter<TokensAndUrlAuthData, MailContainerResource> {
+  private static final long PAGE_SIZE = 50;
+  private static final String USER = "me";
+  private volatile Gmail gmail;
+
+  public GoogleMailExporter() {
+    this.gmail = null;
+  }
+
+  @VisibleForTesting
+  GoogleMailExporter(Gmail gmail) {
+    this.gmail = gmail;
+  }
+
+  @Override
+  public ExportResult<MailContainerResource> export(UUID id, TokensAndUrlAuthData authData) {
+    return export(id, authData, new ExportInformation(null, null));
+  }
+
+  @Override
+  public ExportResult<MailContainerResource> export(UUID id,
+      TokensAndUrlAuthData authData, ExportInformation exportInformation) {
+    // Create a new gmail service for the authorized user
+    Gmail gmail = getOrCreateGmail(authData);
+
+    IdOnlyContainerResource resource =
+        (IdOnlyContainerResource) exportInformation.getContainerResource();
+
+    Messages.List request = null;
```

Comment:
maybe this logic should be moved to a different function? like exportMessages() and exportFolders() ?